src/ol/render/webgl/VectorStyleRenderer.js

```diff
@@ -166,6 +178,31 @@ class VectorStyleRenderer {
       );
     }
 
+    /**
+     * @type {function(import('../../Feature.js').FeatureLike): boolean}
+     * @private
+     */
+    this.featureFilter_ = null;
+    if (filter) {
+      const parsingContext = newParsingContext();
+      const compiled = buildExpression(filter, BooleanType, parsingContext);
```

Comment:
This could be done separately, but I was envisioning that we would also add things like `hasFill`, `hasStroke`, and `hasPoint` (or some other name) to the parsing context.  This would allow us to filter out features based on the style properties being used.  For example, if a style doesn't have any stroke-related properties, then all the linestring geometries could be skipped.

Reply:
I have tried doing this because it sounds like an easy optimization, but after running it though a performance analysis it turns out it doesn't really speed up the render, and on the other hand generates a lot of garbage.

The VectorStyleRenderer uses a mixed geometry batch where geometries are already distributed in three "buckets": polygons, lines and points. If the style doesn't contain any fill instruction, the polygon geometry bucket will simply not be processed. So for this part, filtering beforehand based on geometry type will not bring any improvement.

On the other hand, filtering a geometry batch before processing it takes time because we loop over very feature, and it produces garbage (this might be improved though).

I think I'll leave that out for now, unless you see another opportunity for improving things?